README.md
# Insta Like Button

<a href="https://pub.dev/packages/insta_like_button">
   <img src="https://img.shields.io/pub/v/insta_like_button?color=red" alt="License: MIT"/>
 </a>
 <a href="https://github.com/theiskaa/insta-like-button/blob/main/LICENSE">
   <img src="https://img.shields.io/badge/License-MIT-red.svg" alt="License: MIT"/>
 </a>
 
---

## Installing

### Depend on it

Add this to your package's `pubspec.yaml` file:

```yaml
dependencies:
  insta_like_button: ^0.1.1
```

### Install it

You can install packages from the command line:

```sh
$ flutter pub get
...
```
### Import it

Now in your Dart code, you can use:

```dart
import 'package:insta_like_button/insta_like_button.dart';
```
---
## Usage & Overview
<img src="https://raw.githubusercontent.com/theiskaa/insta-like-button/develop/example/overview/overview.gif" align = "right" height = "500px">

### With required parameters.
```dart
InstaLikeButton(
  image: AssetImage("example/overview/img.jpg"),
  onChanged: () {
    // Do something...
  },
),
```

### With all parameters (Customized InstaLikeButton).
```dart
InstaLikeButton(
  image: NetworkImage("https://picsum.photos/200/300"),
  onChanged: () {
    // Do something...
  },
  icon: Icons.favorite_border,
  iconSize: 80,
  iconColor: Colors.red,
  curve: Curves.fastLinearToSlowEaseIn,
  height: 200,
  width: MediaQuery.of(context).size.width - 20,
  duration: const Duration(seconds: 1),
  onImageError: (e, _) {
    // Do something...
  },
  imageAlignment: Alignment.topLeft,
  imageBoxfit: BoxFit.fill,
  imageScale: 2.0,
  imageColorFilter: ColorFilter.mode(
    Colors.black.withOpacity(0.5),
    BlendMode.dstATop,
  ),
),


CHANGELOG.md
## [0.1.1] - 19/8/2021
- Enabled null safety
- Improved code-quality
- Improved example-quality 

## [0.1.0] - 9/4/2021
- [InstaLikeButton] with these properties: 
- `child`, `onChanged`.
- `duration`, `curve`.
- `icon`, `iconColor`, `iconSize`.
- `height`, `width`.
- `onImageError`, `imageColorFilter`, `imageBoxfit`, `imageAlignment`, `imageScale`.


test/insta_like_button_test.dart
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:insta_like_button/insta_like_button.dart';

void main() {
  InstaLikeButton instaLikeButton;
  late Widget mainWidget;

  setUpAll(() {
    // Initilaze instaLikeButton.
    instaLikeButton = InstaLikeButton(
      image: AssetImage("example/overview/img.jpg"),
      onChanged: () => print("Liked!"),
    );

    // Initilaze mainWidget.
    mainWidget = MaterialApp(
      title: "Insta like button",
      home: Scaffold(body: Center(child: instaLikeButton)),
    );
  });

  testWidgets('Insta like button test', (WidgetTester tester) async {
    await tester.pumpWidget(mainWidget);

    // MainWidget tests.
    expect(find.byType(MaterialApp), findsOneWidget);
    expect(find.byType(Scaffold), findsOneWidget);
    expect(find.byType(Center), findsNWidgets(3));

    // InstaLikeButton tests.
    expect(find.byType(AnimatedBuilder), findsNWidgets(5));
    expect(find.byType(GestureDetector), findsOneWidget);
    expect(find.byType(Container), findsOneWidget);
    expect(find.byType(AnimatedOpacity), findsOneWidget);
  });
}


lib/insta_like_button.dart
library insta_like_button;

import 'dart:async';

import 'package:flutter/material.dart';

/// ### Instagram post style, customizable card widget made for flutter.
///
/// Default Usage: (provided just required parameters).
/// ```dart
/// InstaLikeButton(
///  image: AssetImage("example/overview/img.jpg"),
///  onChanged: () {},
/// );
///
/// ```
/// Other properties which helps to customize [InstaLikeButton].
/// - `duration`, `curve`.
/// - `icon`, `iconColor`, `iconSize`.
/// - `height`, `width`.
/// - `onImageError`, `imageColorFilter`, `imageBoxfit`, `imageAlignment`, `imageScale`.
/// ---
class InstaLikeButton extends StatefulWidget {
  /// This property takes a photo, and builds the entire button on that photo.
  ///
  /// **Usage:**
  ///
  /// add photo by [AssetImage] widget from local path.
  /// ```dart
  /// InstaLikeButton(
  ///  image: AssetImage("assets/photo.png"),
  /// ...
  /// )
  /// ```
  /// or add photo by [NetworkImage] widget, from network.
  /// ```dart
  /// InstaLikeButton(
  ///  image: NetworkImage("https://picsum.photos/200/300"),
  /// ...
  /// )
  /// ```
  final ImageProvider<Object>? image;

  /// The user has tapped the screen with a primary button at the same location
  /// twice in quick succession.
  ///
  /// that's means `icon` will be displayed onto [image] in center.
  /// and would call something you provide within `onChanged` function
  final VoidCallback onChanged;

  /// The icon which would be displayed onto image in center.
  /// as default icon is [Icons.favorite].
  /// If user wanna customize it, then should use it.
  final IconData icon;

  /// The icon's color. as default it equals `Colors.grey[200]`.
  /// If user wanna customize it, then should use it.
  final Color? iconColor;

  /// Height of [InstaLikeButton].
  final double? height;

  /// width of [InstaLikeButton].
  final double? width;

  /// Animation duration. As default it equals [Duration(milliseconds: 500)].
  /// If user wanna customize it, then should use it.
  final Duration? duration;

  /// Curve of like animation. As default it equals [Curves.easeInOut].
  /// If user wanna customize it, then should use it.
  final Curve curve;

  /// size of icon. As default it equals 120
  /// If user wanna customize it, then should use it.
  final double iconSize;

  /// An optional error callback for errors emitted when loading [image].
  final Function(Object, StackTrace?)? onImageError;

  /// A color filter to apply to the image before painting it.
  final ColorFilter? imageColorFilter;

  ///How the [image] should be inscribed into the box.
  final BoxFit? imageBoxfit;

  ///How to align the image within its bounds.
  /// The alignment aligns the given position in the image to the given position in the layout bounds.
  final AlignmentGeometry imageAlignment;

  /// Defines image pixels to be shown per logical pixels.
  ///
  /// By default the value of scale is 1.0.
  /// The scale for the image is calculated by multiplying [scale]
  /// with scale of the given [ImageProvider].
  final double imageScale;

  InstaLikeButton({
    required this.image,
    required this.onChanged,

    ///
    this.icon = Icons.favorite,
    this.iconColor = const Color(0xFFF5F5F5),
    this.height = 300,
    this.width,
    this.duration = const Duration(milliseconds: 500),
    this.curve = Curves.easeInOut,
    this.iconSize = 120,
    this.onImageError,
    this.imageColorFilter,
    this.imageBoxfit,
    this.imageAlignment = Alignment.center,
    this.imageScale = 1.0,
  });
  @override
  _InstaLikeButtonState createState() => _InstaLikeButtonState();
}

class _InstaLikeButtonState extends State<InstaLikeButton>
    with TickerProviderStateMixin {
  late AnimationController _animationController;
  late Animation<double> _iconSizeAnimation;

  // Curve to set animating style of [_iconSizeAnimation].
  late Animation _curve;

  bool _isTapped = false;

  @override
  void dispose() {
    _animationController.dispose();
    super.dispose();
  }

  @override
  void initState() {
    // Initilaze animation controller.
    _animationController = AnimationController(
      vsync: this,
      duration: widget.duration,
    );

    _curve = CurvedAnimation(parent: _animationController, curve: widget.curve);

    _iconSizeAnimation = TweenSequence(
      <TweenSequenceItem<double>>[
        // First act is animating [widget.iconSize - 20] to [widget.iconSize].
        TweenSequenceItem(
          weight: widget.iconSize,
          tween: Tween<double>(
            begin: widget.iconSize - 20,
            end: widget.iconSize,
          ),
        ),

        // Second act is animating [widget.iconSize] to [widget.iconSize - 20].
        TweenSequenceItem(
          weight: widget.iconSize,
          tween: Tween<double>(
            begin: widget.iconSize,
            end: widget.iconSize - 20,
          ),
        ),
      ],

      // Animate with [_curve].
    ).animate(_curve as Animation<double>);

    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return AnimatedBuilder(
      animation: _animationController,
      builder: (context, child) => GestureDetector(
        onDoubleTap: () => onAct(),
        onDoubleTapCancel: () => onActCancel(),
        child: buttonBody(context),
      ),
    );
  }

  // The body of button.
  Container buttonBody(BuildContext context) => Container(
        height: widget.height,
        width: (widget.width != null)
            ? widget.width
            : MediaQuery.of(context).size.width,
        decoration: BoxDecoration(
          image: DecorationImage(
            image: widget.image!,
            onError: widget.onImageError,
            colorFilter: widget.imageColorFilter,
            fit: widget.imageBoxfit,
            scale: widget.imageScale,
            alignment: widget.imageAlignment,
          ),
        ),
        child: Center(child: buildAnimatedIcon()),
      );

  // The animated icon.
  AnimatedOpacity buildAnimatedIcon() => AnimatedOpacity(
        duration: const Duration(milliseconds: 100),
        opacity: _isTapped ? 1 : 0,
        child: Icon(widget.icon,
            size: _iconSizeAnimation.value, color: widget.iconColor),
      );

  // Custom act method which would be used into [onDoubleTap].
  void onAct() {
    // First change [_isTapped] to `true`.
    setState(() => _isTapped = true);

    // Wait 80 milisecond and forward [_animationController]
    Timer(
      const Duration(milliseconds: 80),
      () => _animationController.forward(),
    );

    // Then should call [widget.onTap] function.
    widget.onChanged();

    // Wait one sec.
    // and change [_isTapped] to `false`.
    // and reverse the [_animationController].
    Timer(const Duration(seconds: 1), () {
      setState(() => _isTapped = false);
      _animationController.reverse();
    });
  }

  void onActCancel() => Timer(
        const Duration(seconds: 1),
        () {
          setState(() => _isTapped = false);
          _animationController.reverse();
        },
      );
}


example/main.dart
import 'package:flutter/material.dart';
import 'package:insta_like_button/insta_like_button.dart';

void main() => runApp(App());

class App extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Instagram Like Button/Card',
      home: Home(),
    );
  }
}

class Home extends StatefulWidget {
  @override
  _HomeState createState() => _HomeState();
}

class _HomeState extends State<Home> {
  // To manage the ImageProvider. in error situation.
  var image;

  @override
  Widget build(BuildContext context) {
    // Initilaze [image] when page build.
    image = NetworkImage("https://picsum.photos/200/300");

    return Scaffold(
      body: Center(
        child: Column(
          children: [
            /// With required parameters.
            InstaLikeButton(
              image: AssetImage("example/overview/img.jpg"),
              onChanged: () {
                // Do something...
              },
            ),
            const SizedBox(height: 30),

            /// With all parameters (Customized InstaLikeButton).
            InstaLikeButton(
              image: image,
              onChanged: () {
                // Do something...
              },
              icon: Icons.favorite_border,
              iconSize: 80,
              iconColor: Colors.blue,
              curve: Curves.fastLinearToSlowEaseIn,
              height: 250,
              width: MediaQuery.of(context).size.width - 20,
              duration: const Duration(seconds: 1),
              onImageError: (e, _) => setState(
                () => image = AssetImage("example/overview/img.jpg"),
              ),
              imageAlignment: Alignment.topLeft,
              imageBoxfit: BoxFit.fill,
              imageScale: 2.0,
              imageColorFilter: ColorFilter.mode(
                Colors.black.withOpacity(0.3),
                BlendMode.dstATop,
              ),
            ),
          ],
        ),
      ),
    );
  }
}
